openwillis/measures/text/speech_attribute.py

```diff
@@ -4,154 +4,404 @@
     """
-    text = json_conf['results']['transcripts'][0].get('transcript', '')
-    item_data = json_conf['results']['items']
+    item_data = json_conf["results"]["items"]
+
+    # make a dictionary to map old indices to new indices
+    for i, item in enumerate(item_data):
+        item[measures["old_index"]] = i
+    text = " ".join(
+        [
+            item["alternatives"][0]["content"]
+            for item in item_data
+            if "alternatives" in item
+        ]
+    )
+
+    # phrase-split
+    phrases = nltk.tokenize.sent_tokenize(text)
+    phrases_idxs = []
+
+    start_idx = 0
+    for phrase in phrases:
+        end_idx = start_idx + len(phrase.split()) - 1
+        phrases_idxs.append((start_idx, end_idx))
+        start_idx = end_idx + 1
+
+    # turn-split
+    turns = []
+    turns_idxs = []
+
     if speaker_label is not None:
-        speaker_labels = [item['speaker_label'] for item in item_data if 'speaker_label' in item]
+        speaker_labels = [
+            item["speaker_label"] for item
+            in item_data if "speaker_label" in item
+        ]
 
         if speaker_label not in speaker_labels:
-            raise ValueError(f'Speaker label {speaker_label} not found in the json response object.')
+            raise ValueError(
+                f"Speaker label {speaker_label} "
+                "not found in the json response object."
+            )
+
+        turns_idxs, turns, phrases_idxs, phrases = filter_speaker_label(
+            item_data, speaker_label, turns_idxs, turns, phrases_idxs, phrases
+        )
+
+    # entire transcript - by joining all the phrases
+    text = " ".join(phrases)
+
+    # filter json to only include items with start_time and end_time
+    filter_json = [
+        item for item in item_data
+        if "start_time" in item and "end_time" in item
+    ]
+
+    # calculate time difference between each word
+    for i, item in enumerate(filter_json):
+        if i > 0:
+            item[measures["pause"]] = float(item["start_time"]) - float(
+                filter_json[i - 1]["end_time"]
+            )
+        else:
+            item[measures["pause"]] = np.nan
 
-        # filter the json data based on the speaker label
-        item_data = [item for item in item_data if item.get('speaker_label', '') == speaker_label]
+    if speaker_label is not None:
+        filter_json = [
+            item
+            for item in filter_json
+            if item.get("speaker_label", "") == speaker_label
+        ]
 
-        # extract the text from the filtered json data
-        text_list = [item['alternatives'][0]['content'] for item in item_data if 'alternatives' in item]
-        text = " ".join(text_list)
+    # extract words
+    words = [word["alternatives"][0]["content"] for word in filter_json]
 
-    filter_json = [item for item in item_data if 'start_time' in item and 'end_time' in item]
-    return text, filter_json
+    return (
```

Comment:
too many return values

Suggestion: save all of these in a list and return a single list object, and in speech_characteristics function use the list index to pass them as argument in process_language_feature

Reply:
[group output of filter_transcribe](https://github.com/bklynhlth/openwillis/pull/22/commits/10ee38b7b0caac580e10a1e0b072f6ae49795eea)